Sources/SE0000_KeyPathReflection/Reflection.swift

```diff
@@ -0,0 +1,93 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public enum Reflection {
+  /// Returns the collection of all key paths of this type.
+  ///
+  /// - Parameter type: The static type to return the stored key paths of.
+  /// - Returns: An array of partial key paths for this type.
+  public static func allKeyPaths<T>(for type: T.Type) -> [PartialKeyPath<T>] {
+    guard let metadata = getMetadata(for: type) as? TypeMetadata else {
+      return []
+    }
+    
+    var result = [PartialKeyPath<T>]()
+    result.reserveCapacity(metadata.contextDescriptor.fields.numFields)
+    
+    for i in 0 ..< metadata.contextDescriptor.fields.numFields {
+      let keyPath = createKeyPath(root: metadata, leaf: i) as! PartialKeyPath<T>
+      result.append(keyPath)
+    }
+    
+    return result
+  }
+  
+  /// Returns the collection of all key paths of this value.
+  ///
+  /// - Parameter value: A value of any type to return the stored key paths of.
+  /// - Returns: An array of partial key paths for this value.
+  public static func allKeyPaths<T>(for value: T) -> [PartialKeyPath<T>] {
```

Comment:
@Azoy what do you think about adding versions where you can directly specify the destination type of the key path you want like:
```swift
static func allKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [KeyPath<Source, Destination>] {
    allKeyPaths(from: value).compactMap { kp in
        kp as? KeyPath<Source, Destination>
    }
}
```
And also adding the writable key path version:
```swift
static func allWritableKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [WritableKeyPath<Source, Destination>] {
    allKeyPaths(from: value).compactMap { kp in
        kp as? WritableKeyPath<Source, Destination>
    }
}
```
And finally the recursive versions:
```swift
static func recursivelyAllKeyPaths<Source>(
    for value: Source
) -> [PartialKeyPath<Source>] {
    var result = allKeyPaths(for: value)
    var workList = result
    while let kp = workList.popLast() {
        func helper<T>(_ x: T) {
            let keyPaths = allKeyPaths(for: x).map { kp.appending(path: $0)! }
            result.append(contentsOf: keyPaths)
            workList.append(contentsOf: keyPaths)
        }
        _openExistential(value[keyPath: kp], do: helper)
    }
    return result
}

static func recursivelyAllKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [KeyPath<Source, Destination>] {
    recursivelyAllKeyPaths(for: value).compactMap { kp in
        kp as? KeyPath<Source, Destination>
    }
}

static func recursivelyAllWritableKeyPaths<Source, Destination>(
    from value: Source, to destinationType: Destination.Type
) -> [WritableKeyPath<Source, Destination>] {
    recursivelyAllKeyPaths(for: value).compactMap { kp in
        kp as? WritableKeyPath<Source, Destination>
    }
}
```